lib/src/screens/memory/memory_service.dart

```diff
@@ -32,63 +32,57 @@ class MemoryService {
-          String modelName = eventData[_appEventModelNameKey] ?? 'N/A';
-          String modelData = eventData[_appEventModelDataKey] ?? '{}';
-          String encodedEventData = jsonEncode(EventView(
-            timestamp: event.timestamp!,
-            modelName: modelName,
-            modelData: modelData,
-          ).toJson());
-          memory = WrenchStore.get<Memory>() ?? Memory();
+
+          if (eventData.isNotEmpty &&
+              !eventData.containsKey(_appEventModelNameKey) &&
+              !eventData.containsKey(_appEventModelDataKey)) {
+            continue;
+          }
+
+          String modelName = eventData[_appEventModelNameKey];
+          String modelData = eventData[_appEventModelDataKey];
+
+          AppEvent appEvent = AppEvent().rebuild(
+            (b) => b
+              ..timestamp = event.timestamp ?? 0
+              ..modelName = modelName
+              ..modelData = modelData,
+          );
+
           memory = memory.rebuild(
             (b) => b
-              ..appEvents = memory.appEvents
-                  .rebuild((b) => b.add(encodedEventData))
+              ..appTimelineEvents = memory.appTimelineEvents
+                  .rebuild((b) => b.add(appEvent))
                   .toBuilder()
               ..appState = memory.appState
-                  .rebuild((b) => b.updateValue(
-                      modelName, (_) => encodedEventData,
-                      ifAbsent: () => encodedEventData))
+                  .rebuild((b) => b.updateValue(modelName, (_) => appEvent,
+                      ifAbsent: () => appEvent))
                   .toBuilder()
+              ..modelData = modelData
               ..scroll = true,
           );
 
-          List<String> targetEvents = memory.appEvents.toList();
-          List<String> searchedTargetEvent = [];
-          for (String event in targetEvents) {
-            EventView eventView = EventView.fromJson(jsonDecode(event));
-            if (eventView.modelName
-                .toLowerCase()
-                .contains(memory.modelDataSearchText.toLowerCase())) {
-              searchedTargetEvent.add(event);
-            }
-          }
-          String eventDataString =
-              EventView.fromJson(jsonDecode(searchedTargetEvent.last))
-                  .modelData;
-          int indexOfLastEvent = (searchedTargetEvent.length - 1);
+          List<AppEvent> filteredTimelineEvents = memory.appTimelineEvents
+              .where((e) => e.modelName
+                  .toLowerCase()
+                  .contains(memory.timelineModelSearchText.toLowerCase()))
+              .toList();
+
           memory = memory.rebuild(
             (b) => b
-              ..modelData = eventDataString
-              ..filteredAppEvents = ListBuilder<String>(searchedTargetEvent)
-              ..selectedAppEventIndex = indexOfLastEvent,
+              ..filteredAppTimelineEvents = filteredTimelineEvents.isNotEmpty
```

Comment:
unless you want to show the entire timeline even when the user typed model name doesn't exist

Reply:
I personally feel we should show `No model found` message instead. but this works too

Reply:
Type some string in timeline filter field, search works as expected. Now clear the string, input is now `''`, then `filteredTimelineEvents` will be empty becaues `contains` is looking for empty string. Instead of showing all models, UI shows current contents of `filteredTimelineEvents`